File: scase-pulsar/src/main/scala/io/jobial/scase/pulsar/PulsarMessageHandlerServiceConfiguration.scala
package io.jobial.scase.pulsar

import cats.effect.{Concurrent, ContextShift, IO}
import cats.implicits._
import io.jobial.scase.core.impl.ConsumerMessageHandlerService
import io.jobial.scase.core.{MessageHandler, ServiceConfiguration}
import io.jobial.scase.marshalling.Unmarshaller

case class PulsarMessageHandlerServiceConfiguration[M: Unmarshaller](
  serviceName: String,
  topic: String
) extends ServiceConfiguration {

  def service[F[_] : Concurrent](handler: MessageHandler[F, M])
    (
      implicit context: PulsarContext,
      cs: ContextShift[IO]
    ) =
    for {
      consumer <- PulsarConsumer[F, M](topic)
      service = ConsumerMessageHandlerService(
        consumer,
        handler
      )
    } yield service

}

File: scase-pulsar/src/main/java/io/jobial/scase/pulsar/javadsl/PulsarMessageHandlerServiceConfiguration.java
package io.jobial.scase.pulsar.javadsl;

import cats.effect.IO;
import io.jobial.scase.core.MessageHandler;
import io.jobial.scase.core.javadsl.JavaUtils;
import io.jobial.scase.core.javadsl.SenderClient;
import io.jobial.scase.core.javadsl.Service;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import static io.jobial.scase.core.javadsl.JavaUtils.ioAsyncEffect;

public class PulsarMessageHandlerServiceConfiguration<M> {

    io.jobial.scase.pulsar.PulsarMessageHandlerServiceConfiguration config;

    PulsarMessageHandlerServiceConfiguration(io.jobial.scase.pulsar.PulsarMessageHandlerServiceConfiguration<M> config) {
        this.config = config;
    }

    public CompletableFuture<Service> service(MessageHandler<IO, M> messageHandler, PulsarContext pulsarContext) throws ExecutionException, InterruptedException {
        return JavaUtils.service(config.service(messageHandler, ioAsyncEffect, pulsarContext.getContext()));
    }

    public CompletableFuture<Service> service(MessageHandler<IO, M> messageHandler) throws ExecutionException, InterruptedException {
        return service(messageHandler, new PulsarContext());
    }

    public CompletableFuture<SenderClient<M>> client(PulsarContext pulsarContext) throws ExecutionException, InterruptedException {
        return JavaUtils.senderClient(config.client(ioAsyncEffect, pulsarContext.getContext()));
    }

    public CompletableFuture<SenderClient<M>> client() throws ExecutionException, InterruptedException {
        return client(new PulsarContext());
    }
}
